File: backend/app/api/swings.py
import json
import logging
import os

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, UploadFile
from fastapi.responses import Response
from sqlalchemy.orm import Session

from app.api.auth import get_current_user_id
from app.db.session import get_db
from app.models.swing import Swing
from app.pipeline.orchestrator import analyze_swing
from app.schemas.swing import SwingListResponse, SwingSummary, SwingUploadResponse
from app.storage.video_store import save_video

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/swings", tags=["swings"])


def _process_swing(swing_id: str, db_url: str):
    from sqlalchemy import create_engine
    from sqlalchemy.orm import Session as SA_Session
    from sqlalchemy.orm import sessionmaker

    engine = create_engine(db_url)
    SessionLocal = sessionmaker(bind=engine)
    db = SessionLocal()

    try:
        swing = db.query(Swing).filter(Swing.id == swing_id).first()
        if not swing:
            return

        swing.status = "processing"
        db.commit()

        result = analyze_swing(
            video_path=swing.video_path,
            user_id=swing.user_id,
            swing_id=swing.id,
            handedness=swing.handedness,
        )

        swing.status = "complete"
        swing.overall_score = result.overall_score
        swing.analysis_json = result.model_dump_json()
        db.commit()
    except Exception as e:
        logger.exception(f"Failed to process swing {swing_id}")
        swing = db.query(Swing).filter(Swing.id == swing_id).first()
        if swing:
            swing.status = "failed"
            swing.error_message = str(e)[:500]
            db.commit()
    finally:
        db.close()
        engine.dispose()


@router.post("/upload", response_model=SwingUploadResponse, status_code=202)
async def upload_swing(
    video: UploadFile,
    background_tasks: BackgroundTasks,
    handedness: str = "right",
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    if video.content_type and not video.content_type.startswith("video/"):
        raise HTTPException(status_code=400, detail="File must be a video")

    data = await video.read()
    ext = os.path.splitext(video.filename or ".mp4")[1] or ".mp4"
    video_path = save_video(data, extension=ext)

    swing = Swing(
        user_id=user_id,
        video_path=video_path,
        handedness=handedness,
        status="pending",
    )
    db.add(swing)
    db.commit()
    db.refresh(swing)

    from app.config import settings

    background_tasks.add_task(_process_swing, swing.id, settings.database_url)

    return SwingUploadResponse(swing_id=swing.id, status="processing")


@router.get("/{swing_id}")
def get_swing(
    swing_id: str,
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    swing = (
        db.query(Swing)
        .filter(Swing.id == swing_id, Swing.user_id == user_id)
        .first()
    )
    if not swing:
        raise HTTPException(status_code=404, detail="Swing not found")

    if swing.status != "complete" or not swing.analysis_json:
        return {
            "swing_id": swing.id,
            "status": swing.status,
            "error_message": swing.error_message,
        }

    return json.loads(swing.analysis_json)


@router.get("")
def list_swings(
    page: int = 1,
    page_size: int = 20,
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    query = db.query(Swing).filter(Swing.user_id == user_id)
    total = query.count()

    swings = (
        query.order_by(Swing.created_at.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
        .all()
    )

    return SwingListResponse(
        items=[SwingSummary.model_validate(s) for s in swings],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.get("/{swing_id}/frames/{phase}")
def get_swing_frame(
    swing_id: str,
    phase: str,
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    """Get a key frame with skeleton overlay for a specific phase."""
    swing = (
        db.query(Swing)
        .filter(Swing.id == swing_id, Swing.user_id == user_id)
        .first()
    )
    if not swing or swing.status != "complete" or not swing.analysis_json:
        raise HTTPException(status_code=404, detail="Swing not found or not complete")

    analysis = json.loads(swing.analysis_json)

    # Find the key frame for the requested phase (midpoint of phase range)
    phase_data = None
    for p in analysis.get("phases_detected", []):
        if p["phase"] == phase:
            phase_data = p
            break

    if not phase_data:
        raise HTTPException(status_code=404, detail=f"Phase '{phase}' not found")

    key_frame_idx = (phase_data["start_frame"] + phase_data["end_frame"]) // 2

    from app.pipeline.frame_renderer import render_key_frames

    frames = render_key_frames(
        video_path=swing.video_path,
        pose_frames_data=analysis.get("pose_frames", []),
        phase_key_frames={phase: key_frame_idx},
    )

    jpeg_bytes = frames.get(phase)
    if not jpeg_bytes:
        raise HTTPException(status_code=404, detail="Frame not available")

    return Response(content=jpeg_bytes, media_type="image/jpeg")


@router.get("/{swing_id}/frames")
def list_swing_frames(
    swing_id: str,
    user_id: str = Depends(get_current_user_id),
    db: Session = Depends(get_db),
):
    """List available phase frames for a swing."""
    swing = (
        db.query(Swing)
        .filter(Swing.id == swing_id, Swing.user_id == user_id)
        .first()
    )
    if not swing or swing.status != "complete" or not swing.analysis_json:
        raise HTTPException(status_code=404, detail="Swing not found or not complete")

    analysis = json.loads(swing.analysis_json)
    phases = [p["phase"] for p in analysis.get("phases_detected", [])]

    from app.config import settings

    base_url = f"/api/swings/{swing_id}/frames"
    return {
        "swing_id": swing_id,
        "frames": [
            {"phase": phase, "url": f"{base_url}/{phase}"}
            for phase in phases
        ],
    }

File: api/internal/handler/swings.go
package handler

import (
	"encoding/json"
	"fmt"
	"path/filepath"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
	"gorm.io/gorm"

	"github.com/6cclab/swing-detector/api/internal/middleware"
	"github.com/6cclab/swing-detector/api/internal/models"
	"github.com/6cclab/swing-detector/api/internal/queue"
	"github.com/6cclab/swing-detector/api/internal/storage"
)

type SwingHandler struct {
	db    *gorm.DB
	store storage.Storage
	queue *queue.Queue
}

func NewSwingHandler(db *gorm.DB, store storage.Storage, q *queue.Queue) *SwingHandler {
	return &SwingHandler{db: db, store: store, queue: q}
}

func (h *SwingHandler) Upload(c *fiber.Ctx) error {
	userID := middleware.GetUserID(c)

	file, err := c.FormFile("video")
	if err != nil {
		return c.Status(400).JSON(fiber.Map{"error": "Video file is required"})
	}

	if ct := file.Header.Get("Content-Type"); ct != "" {
		if len(ct) < 6 || ct[:6] != "video/" {
			return c.Status(400).JSON(fiber.Map{"error": "File must be a video"})
		}
	}

	handedness := c.Query("handedness", "right")

	f, err := file.Open()
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to read file"})
	}
	defer f.Close()

	ext := filepath.Ext(file.Filename)
	if ext == "" {
		ext = ".mp4"
	}

	storedPath, err := h.store.SaveVideoStream(f, ext)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to save video"})
	}

	swingID := uuid.New().String()
	swing := models.Swing{
		ID:         swingID,
		UserID:     userID,
		VideoPath:  storedPath,
		Handedness: handedness,
		Status:     "pending",
	}

	if err := h.db.Create(&swing).Error; err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to create swing record"})
	}

	if err := h.queue.Publish(c.Context(), queue.SwingJob{
		SwingID:    swingID,
		VideoPath:  storedPath,
		Handedness: handedness,
		UserID:     userID,
	}); err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to queue analysis"})
	}

	return c.Status(202).JSON(models.SwingUploadResponse{
		SwingID: swingID,
		Status:  "processing",
	})
}

func (h *SwingHandler) Get(c *fiber.Ctx) error {
	userID := middleware.GetUserID(c)
	swingID := c.Params("id")

	var swing models.Swing
	if err := h.db.Where("id = ? AND user_id = ?", swingID, userID).First(&swing).Error; err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "Swing not found"})
	}

	if swing.Status != "complete" || swing.AnalysisJSON == nil {
		return c.JSON(fiber.Map{
			"swing_id":      swing.ID,
			"status":        swing.Status,
			"error_message": swing.ErrorMessage,
		})
	}

	var result json.RawMessage
	if err := json.Unmarshal([]byte(*swing.AnalysisJSON), &result); err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to parse analysis"})
	}

	return c.JSON(result)
}

func (h *SwingHandler) List(c *fiber.Ctx) error {
	userID := middleware.GetUserID(c)
	page := c.QueryInt("page", 1)
	pageSize := c.QueryInt("page_size", 20)

	if page < 1 {
		page = 1
	}

	var total int64
	h.db.Model(&models.Swing{}).Where("user_id = ?", userID).Count(&total)

	var swings []models.Swing
	h.db.Where("user_id = ?", userID).
		Order("created_at DESC").
		Offset((page - 1) * pageSize).
		Limit(pageSize).
		Find(&swings)

	items := make([]models.SwingSummary, len(swings))
	for i, s := range swings {
		items[i] = models.SwingSummary{
			ID:           s.ID,
			CreatedAt:    s.CreatedAt,
			Status:       s.Status,
			OverallScore: s.OverallScore,
			Handedness:   s.Handedness,
		}
	}

	return c.JSON(models.SwingListResponse{
		Items:    items,
		Total:    total,
		Page:     page,
		PageSize: pageSize,
	})
}

func (h *SwingHandler) ListFrames(c *fiber.Ctx) error {
	userID := middleware.GetUserID(c)
	swingID := c.Params("id")

	var swing models.Swing
	if err := h.db.Where("id = ? AND user_id = ?", swingID, userID).First(&swing).Error; err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "Swing not found"})
	}

	if swing.Status != "complete" || swing.AnalysisJSON == nil {
		return c.Status(404).JSON(fiber.Map{"error": "Swing not complete"})
	}

	var analysis struct {
		PhasesDetected []struct {
			Phase string `json:"phase"`
		} `json:"phases_detected"`
	}
	if err := json.Unmarshal([]byte(*swing.AnalysisJSON), &analysis); err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to parse analysis"})
	}

	type frameInfo struct {
		Phase string `json:"phase"`
		URL   string `json:"url"`
	}

	frames := make([]frameInfo, len(analysis.PhasesDetected))
	for i, p := range analysis.PhasesDetected {
		frames[i] = frameInfo{
			Phase: p.Phase,
			URL:   fmt.Sprintf("/api/swings/%s/frames/%s", swingID, p.Phase),
		}
	}

	return c.JSON(fiber.Map{
		"swing_id": swingID,
		"frames":   frames,
	})
}

func (h *SwingHandler) GetFrame(c *fiber.Ctx) error {
	userID := middleware.GetUserID(c)
	swingID := c.Params("id")
	phase := c.Params("phase")

	var swing models.Swing
	if err := h.db.Where("id = ? AND user_id = ?", swingID, userID).First(&swing).Error; err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "Swing not found"})
	}

	if swing.Status != "complete" {
		return c.Status(404).JSON(fiber.Map{"error": "Swing not complete"})
	}

	data, err := h.store.ReadFrame(swingID, phase)
	if err != nil {
		return c.Status(404).JSON(fiber.Map{"error": "Frame not available"})
	}

	c.Set("Content-Type", "image/jpeg")
	return c.Send(data)
}
